Jetpack Social: Add an image picker to the Jetpack Social editor panel

danielpost opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
This PR adds an image picker to the Jetpack Social panel when the enhanced publishing feature is enabled.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

These steps are identical for Jetpack and Jetpack Social.
* Start writing a new post and check that the Jetpack Social panel looks the same as it did before.
* Add support for enhanced publishing to your site (context here: D90213-code)
* Refresh the page. You should now see an image picker (see screenshot).
* Pick an image, and make sure that you can delete and change it properly.
* Save as **draft**, refresh the page and check that the image got saved properly.
<img width="285" alt="image" src="https://user-images.githubusercontent.com/1713699/198008294-e1f73cbd-69d8-4847-ab40-4d9d1aab65ad.png">

Commit:
Add media section for Jetpack Social

Commit:
changelog

Commit:
Update dependencies

Commit:
Save attached media to post meta

Commit:
changelog

Commit:
Merge branch 'trunk' into add/jetpack-social-panel-save-attached-media

Commit:
Fix README for useAttachedMedia hook

Commit:
Update dependencies

Commit:
Redesign media selector to match core featured image

Commit:
Merge branch 'trunk' into add/jetpack-social-panel-save-attached-media

Commit:
Add feature check

Commit:
Update dependencies

Commit:
Remove redundant changelogs

danielpost commented on projects/js-packages/publicize-components/src/components/media-section/index.js:
```diff
@@ -0,0 +1,100 @@
+	const onRemoveMedia = useCallback( () => updateAttachedMedia( [] ), [ updateAttachedMedia ] );
+	const onUpdateMedia = useCallback( media => updateAttachedMedia( [ media.id ] ), [
+		updateAttachedMedia,
+	] );
```
I've used `useCallback` here because ESLint complains about inline arrow functions without it.

danielpost commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1444,39 +1469,75 @@ public function publicize_connections_url( $source = 'calypso-marketing-connecti
 	 */
+	public function get_api_data( $blog_id ) {
+		static $api_data_response = null;
+		$key                      = 'jetpack_social_api_data';
```
I've updated the key of this transient, which will invalidate all existing ones. However, it shouldn't be an issue considering this methods makes an API call every time it gets called anyway, and only uses the transient in case of an error.

pablinos replied:
Makes sense. I'm glad you're caching the response for the request too.

Commit:
Update readme file

Commit:
Fix typedoc and check

danielpost commented on projects/packages/sync/src/class-defaults.php:
```diff
@@ -723,6 +723,7 @@ public static function get_multisite_callable_whitelist() {
 		'_wpas_feature_enabled',
 		'_wpas_is_tweetstorm',
 		'_wpas_mess',
+		'_wpas_attached_media',
```
I'll add it to WPCOM separately (see PCYsg-sBM-p2)

pablinos replied:
Yes, I've been starting to wonder if we should swap to a single serialised value to prevent the need for this in future.

gmjuhasz commented on projects/js-packages/publicize-components/src/components/media-section/index.js:
```diff
@@ -0,0 +1,106 @@
+					render={ setMediaRender }
+					value={ attachedMedia[ 0 ] }
+				/>
+				{ mediaObject && (
```
I was wondering if we need this replace button. Clicking on the image itself brings up the image picker too. The primary button with the borders and the basic remove link below it looks a bit strange together in my opinion, but maybe I am wrong!

pablinos replied:
I agree, I think we should be able to remove the replace button.

pablinos replied:
I see now that the featured image has a replace button too. Perhaps this is for accessibility?

gmjuhasz reviewed (commented):
Nice work on this, it looks great! I just have a few things in mind, and a minor issue

* It is possible to bypass the allowed media types, by uploading a GIF for example. I can even upload a ZIP! It does not appear in the media library, but after uploading it I can select it, and it gets loaded into the editor. This could create unwanted issues when talking with the API later.

* I have some weirdness going on with the images. Some images are not scaled down to the size of the box, and it goes out from it hiding other stuff. It's weird because for this particular image the retrieved thumbnail size is `150x150` which is not in the image's ratio, so I think that's why it can't rescale it.

<img width="287" alt="CleanShot 2022-10-27 at 10 56 12@2x" src="https://user-images.githubusercontent.com/36671565/198240404-cd52c2f2-d6aa-4bfa-8802-28c36c5a28c2.png">

Commit:
Fix image ratio

danielpost commented:
@gmjuhasz Thanks!

> It is possible to bypass the allowed media types, by uploading a GIF for example. I can even upload a ZIP! It does not appear in the media library, but after uploading it I can select it, and it gets loaded into the editor. This could create unwanted issues when talking with the API later.

I can reproduce this, however I can also reproduce this with the core Featured Image component. It appears to be a core bug in Gutenberg: https://github.com/WordPress/gutenberg/issues/25130.

I've added a manual MIME check. Not ideal but I don't think there's any way around it at the moment.

> I have some weirdness going on with the images. Some images are not scaled down to the size of the box, and it goes out from it hiding other stuff. It's weird because for this particular image the retrieved thumbnail size is 150x150 which is not in the image's ratio, so I think that's why it can't rescale it.

Fixed! Was missing some CSS.

> I was wondering if we need this replace button. Clicking on the image itself brings up the image picker too. The primary button with the borders and the basic remove link below it looks a bit strange together in my opinion, but maybe I am wrong!

I don't disagree, but the core Featured Image component has the same layout and I think it would be smart to stick to that as much as possible.

gmjuhasz commented:
> I've added a manual MIME check. Not ideal but I don't think there's any way around it at the moment.

Sounds good for now, hopefully it will get fixed soon.

>I don't disagree, but the core Featured Image component has the same layout and I think it would be smart to stick to that as much as possible.

Yeah, I didn't know that was the case, makes sense then!

I think it looks good, the image scaling worked too!

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1444,39 +1469,75 @@ public function publicize_connections_url( $source = 'calypso-marketing-connecti
+			return false;
 		}
 
-		return ! is_wp_error( $share_info_response ) ? $share_info_response : null;
```
I think how this has been refactored we're returning an empty array rather than `null`. It shouldn't be an issue, but it might be worth checking.

danielpost replied:
The new `get_api_data` method always returns an (empty) array, but the existing `get_publicize_shares_info` still returns either an array or `null`.

danielpost replied:
I've also swapped to a single serialized value in [`71ccfea`](https://github.com/Automattic/jetpack/pull/26930/commits/71ccfea9ab1580ba3c8494b6a041cf7e6e3fc894).

pablinos commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -712,8 +712,9 @@ public static function enqueue_block_editor_assets() {
 			$initial_state['social'] = array(
-				'sharesData'  => $publicize->get_publicize_shares_info( $blog_id ),
-				'hasPaidPlan' => $publicize->has_paid_plan(),
+				'sharesData'                  => $publicize->get_publicize_shares_info( $blog_id ),
+				'hasPaidPlan'                 => $publicize->has_paid_plan(),
+				'isEnhancedPublishingEnabled' => $publicize->is_enhanced_publishing_enabled( $blog_id ),
```
Probably not for now, but I wonder as this is all coming from the same API call, if we could consolidate this to `$publicize->get_initial_state()` or something.

danielpost replied:
Agreed, I'll add it to our list 👍🏼

pablinos commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -11,9 +11,11 @@ import { getSiteFragment } from '@automattic/jetpack-shared-extension-utils';
 import { PanelRow, Disabled, ExternalLink } from '@wordpress/components';
 import { Fragment, createInterpolateElement } from '@wordpress/element';
 import { _n, sprintf } from '@wordpress/i18n';
+import usePublicizeConfig from '../../hooks/use-publicize-config';
```
For some reason we are passing in the other values supplied from this hook as props to the form component. Shall we pass in `isEnhancedPublishingEnabled` too or swap to getting the other values from `usePublicizeConfig`?

danielpost replied:
We can swap to getting the other values from the hook in a follow-up? I'm not sure why it's set up the current way so it's worth investigating a bit more, but I feel like the whole point of having the hook is to avoid having to pass down those props.

pablinos replied:
Let's do it one way or the other. That's the kind of follow up that we won't get to!

Commit:
Add mime check

pablinos commented on projects/js-packages/publicize-components/src/components/media-section/styles.module.scss:
```diff
@@ -0,0 +1,61 @@
+		margin-top: 1em;
+	}
+
+	// This keeps images at their intrinsic size (eg. a 50px
+	// image will never be wider than 50px).
+	& :global( .components-responsive-wrapper__content ) {
+		max-width: 100%;
+		width: auto;
+		height: auto;
+	}
```
Why do we need to do this? If there is an issue with the image picker component, can we fix it upstream? We shouldn't have small images here anyway. We could maybe deal with that when we work on validation.

pablinos replied:
Ok, I see this is for the issue that @gmjuhasz. It's actually just the `width: auto` which fixes it. I feel like this is something that should be fixed in the component. I see that the featured image is patching it in a similar way!

danielpost replied:
Yeah, this all comes from the core featured image component.

pablinos replied:
Yeah, we can reduce this to just `width: auto`, or at least that was working for me. I feel like we should open a PR on Gutenberg to fix this in the more general case.

pablinos commented on projects/js-packages/publicize-components/src/components/media-section/styles.module.scss:
```diff
@@ -0,0 +1,61 @@
+.wrapper {
+	padding: 0;
+
+	& :global( .components-spinner ) {
+		position: absolute;
+		top: 50%;
+		left: 50%;
+		margin-top: -9px;
+		margin-left: -9px;
+	}
```
Similarly is there a reason why we need to adjust the spinner of the core component? This seems like a change that should be handled upstream.

danielpost replied:
It's more of a mixing of core and custom components, we're using the core spinner component in our custom media picker component, which is why it needs some extra styling.

pablinos commented on projects/js-packages/publicize-components/src/components/media-section/styles.module.scss:
```diff
@@ -0,0 +1,61 @@
+		margin-left: -9px;
+	}
+
+	// Stack consecutive buttons.
+	& :global( .components-button + .components-button ) {
+		display: block;
+		margin-top: 1em;
+	}
```
If we remove the replace button then do you think we could remove this?

pablinos commented on projects/js-packages/publicize-components/src/hooks/use-attached-media/index.js:
```diff
@@ -0,0 +1,30 @@
+
+	return {
+		attachedMedia,
+		updateAttachedMedia: function ( ids ) {
+			editPost( {
+				meta: {
+					jetpack_publicize_attached_media: ids,
+				},
+			} );
+		},
```
I wonder if this should be wrapped in `useCallback`? I don't know for sure the performance implications, but `updateAttachedMedia` will change on each call to the hook, and that will invalidate the `useCallback` calls, that have it in its dependencies.

Maybe we could implement dependencies for this hook?

danielpost replied:
Can you elaborate a bit on this? I think it makes sense in theory, but I'm not sure exactly where to add the `useCallback`. Do you mean wrapping the `editPost` in a `useCallback`?

pablinos commented on projects/js-packages/publicize-components/src/components/media-section/index.js:
```diff
@@ -0,0 +1,115 @@
+ * @returns {object} The media section.
+ */
+export default function MediaSection() {
+	const ADD_MEDIA_LABEL = __( 'Set Social Image', 'jetpack' );
+	const REPLACE_MEDIA_LABEL = __( 'Replace Social Image', 'jetpack' );
+	const REMOVE_MEDIA_LABEL = __( 'Remove Social Image', 'jetpack' );
+
+	const ALLOWED_MEDIA_TYPES = useMemo( () => [ 'image/jpeg', 'image/png' ], [] );
```
If we were to define these outside of `MediaSection` like `getMediaDetails` that should mean we can avoid having to memoize ALLOWED_MEDIA_TYPES, and subsequently adding these to hook dependency arrays.

danielpost replied:
Good idea! Fixed.

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1050,6 +1074,7 @@ public function register_post_meta() {
 			register_meta( 'post', $this->POST_MESS, $message_args );
 			register_meta( 'post', $this->POST_TWEETSTORM, $tweetstorm_args );
 			register_meta( 'post', self::POST_PUBLICIZE_FEATURE_ENABLED, $publicize_feature_enable_args );
+			register_meta( 'post', self::POST_ATTACHED_MEDIA, $publicize_attached_media_args );
```
I've just realised that we should be registering these meta values for any post type that supports Publicize. Maybe not tweetstorm as that's a bit specific to posts, but the other values would be needed.

danielpost replied:
`post` should cover all custom post types as well—you can specify `object_subtype` to add it just for a specific post type: https://developer.wordpress.org/reference/functions/register_meta/

danielpost replied:
We're actually looping through all supported post types explicitly already: https://github.com/Automattic/jetpack/blob/trunk/projects/packages/publicize/src/class-publicize-base.php#L1041

pablinos reviewed (commented):
It's looking good! I can attach an image, and it gets saved to the meta correctly. 

There's the usual array of thoughts and comments, a number of which are bigger changes that are for follow-ups if we do them at all 😄

Commit:
[not verified] Merge branch 'trunk' into add/jetpack-social-panel-save-attached-media

Commit:
Switch to serialized option

Commit:
Remove unnecessary useMemo checks

Commit:
Update Jetpack Social options meta key

Commit:
Remove unnecessary CSS, fix media method to support odd-sized images

Commit:
Reinstate Replace Image button

Commit:
Move isEnhancedPublishingEnabled to rest of config

Commit:
Merge branch 'trunk' into add/jetpack-social-panel-save-attached-media

pablinos commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php:
```diff
@@ -23,6 +23,7 @@
  *   meta: { # Not defined in this file. Handled in modules/publicize/publicize.php via `register_meta()`
  *     jetpack_publicize_feature_enabled: (boolean) Is this publicize feature enabled?
  *     jetpack_publicize_message: (string) The message to use instead of the post's title when sharing.
+ *     jetpack_publicize_attached_media: (array) List of media IDs that will be attached to the social media post.
```
Is this true? I think it will be the `jetpack_social_options` now.

There's [a second version of this file](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/publicize/src/class-connections-post-field.php) in the publicize package too

danielpost replied:
Good catch! Updated in [`dd96f82`](https://github.com/Automattic/jetpack/pull/26930/commits/dd96f820c660f6b2bd7b2070e99c83f594822b10).

I don't think that other file needs to be updated (other than maybe the PHPDoc for all meta?).

Commit:
Update dependencies

pablinos commented on projects/js-packages/publicize-components/src/hooks/use-attached-media/index.js:
```diff
@@ -0,0 +1,33 @@
+
+	return {
+		attachedMedia,
+		updateAttachedMedia: function ( ids ) {
```
As this is essentially an arrow function, do you think there's an advantage in wrapping this in a `useCallback`? We could also put `currentOptions` within the function, so it isn't a dependency and we only call `getJetpackSocialOptions` when we need to.

pablinos replied:
I've been researching this a bit more, and came across [my favourite SO answer on it](https://stackoverflow.com/a/64134704/5923608). I think we probably do want to wrap this function in `useCallback` so `updateAttachedMedia` is stable between renders, otherwise it will invalidate the `useCallbacks` where it's being used.

danielpost replied:
Makes sense! Fixed in [`0c86a13`](https://github.com/Automattic/jetpack/pull/26930/commits/0c86a13512703d5c596af9f5b83619c7dba77e52).

pablinos reviewed (commented):
This is working well for me. Just a couple of minor comments. Sorry if I missed the hook thing before. I'm not certain about the performance of it, but it seems like it might be a good idea, so the callbacks aren't recreated on each render.

Commit:
Wrap updateAttachedMedia in useCallback

Commit:
Update docs for post meta

Commit:
Merge branch 'trunk' into add/jetpack-social-panel-save-attached-media

Commit:
Update dependencies

spsiddarthan reviewed (approved):
This tests well per the instructions and I see the other comments have been addressed as well, @danielpost!
